sno/conflicts.py

```diff
@@ -12,6 +14,113 @@
 L = logging.getLogger("sno.conflicts")
 
 
+class ConflictIndex:
+    """
+    Like a pygit2.Index, but every conflict has a short key independent of its path,
+    and the entire index including conflicts can be serialised to a tree.
+    Conflicts are easier to modify than in a pygit2.Index (where they are backed by C iterators).
+    When serialised to a tree, conflicts will be added in a special .conflicts/ directory.
+    """
+
+    def __init__(self, index, serialised=False):
+        self.entries = {}
+        self.conflicts = {}
+
+        if serialised and index.conflicts:
+            raise RuntimeError(
+                "pygit2.Index.conflicts should be empty if index has been serialised"
+            )
+
+        for entry in index:
+            if entry.path.startswith(".conflicts/"):
+                if not serialised:
+                    raise RuntimeError(
+                        ".conflicts/ directory shouldn't exist if index has not been serialised"
+                    )
+                key, conflict_part = self._deserialise_conflict_part(entry)
+                self.conflicts.setdefault(key, AncestorOursTheirs.EMPTY)
+                self.add_conflict(key, self.conflicts.get(key) | conflict_part)
+            else:
+                self.add(entry)
+
+        if index.conflicts:
+            for key, conflict3 in enumerate(index.conflicts):
+                self.add_conflict(str(key), conflict3)
+
+    def add(self, index_entry):
+        assert type(index_entry) == pygit2.IndexEntry
+        self.entries[index_entry.path] = index_entry
+
+    def remove(self, path):
+        del self.entries[path]
+
+    def __iter__(self):
+        return iter(self.entries.values())
+
+    def __getitem__(self, path):
+        return self.entries[path]
+
+    def __setitem__(self, path, index_entry):
+        assert path == index_entry.path
+        self.entries[path] = index_entry
+
+    def add_conflict(self, key, conflict3):
+        assert type(key) == str
+        self.conflicts[key] = AncestorOursTheirs.of(conflict3)
+
+    def remove_conflict(self, key):
+        del self.conflicts[key]
+
+    def _serialise_conflict(self, key, conflict):
+        result = []
+        for version, index_entry in zip(AncestorOursTheirs.NAMES, conflict):
+            if not index_entry:
+                continue
+            result_path = f".conflicts/{key}/{version}/{index_entry.path}"
+            result.append(
+                pygit2.IndexEntry(result_path, index_entry.id, index_entry.mode)
+            )
+        return result
+
+    _PATTERN = re.compile(
+        r"^.conflicts/(?P<key>.*)/(?P<version>ancestor|ours|theirs)/(?P<path>.*)$"
```

Comment:
```suggestion
        r"^\.conflicts/(?P<key>.+?)/(?P<version>ancestor|ours|theirs)/(?P<path>.+)$"
```

The `?` in the `key` match is important to make it non-greedy, otherwise if `path` contains `"ancestor"` I think `key` would gobble up too much

Reply:
Done